packages/nimble-components/src/wafer-map/modules/rendering.ts

```diff
@@ -5,41 +5,72 @@ import type { DataManager } from './data-manager';
  * Responsible for drawing the dies inside the wafer map
  */
 export class RenderingModule {
-    private readonly waferData: DataManager;
     private readonly context: CanvasRenderingContext2D;
+    private dieSize: number | undefined;
+    private readonly dies: DieRenderInfo[];
+    private readonly dimensions: Dimensions;
+    private readonly labelFontSize: number;
 
     public constructor(waferData: DataManager, canvas: HTMLCanvasElement) {
-        this.waferData = waferData;
         this.context = canvas.getContext('2d')!;
+        this.dies = waferData.diesRenderInfo;
+        this.dimensions = waferData.dieDimensions;
+        this.labelFontSize = waferData.labelsFontSize;
     }
 
-    public drawWafer(): void {
-        const dies: DieRenderInfo[] = this.waferData.diesRenderInfo;
-        const dimensions: Dimensions = this.waferData.dieDimensions;
+    public drawWafer(transform?: number): void {
+        this.renderDies(transform);
+        this.renderText(transform);
+    }
 
-        for (const die of dies) {
-            this.context.fillStyle = die.fillStyle;
-            this.context?.fillRect(
-                die.x,
-                die.y,
-                dimensions.width,
-                dimensions.height
-            );
+    public clearCanvas(width: number, height: number): void {
+        this.context.clearRect(0, 0, width, height);
+    }
 
-            this.context.font = this.waferData.labelsFontSize.toString();
-            this.context.fillStyle = '#ffffff';
-            this.context.textAlign = 'center';
-            const aproxTextHeight = this.context.measureText('M');
+    private renderDies(transform?: number): void {
+        this.dieSize = this.dimensions.width * this.dimensions.height * (transform || 1);
+        this.dies.sort((a, b) => a.fillStyle.localeCompare(b.fillStyle));
+        let prev!: DieRenderInfo;
 
-            this.context.fillText(
-                die.text,
-                die.x + dimensions.width / 2,
-                die.y + dimensions.height / 2 + aproxTextHeight.width / 2
+        for (const die of this.dies) {
+            if (!prev) {
+                this.context.fillStyle = die.fillStyle;
+            }
+            if (prev && die.fillStyle !== prev.fillStyle && die.fillStyle) {
+                this.context.fillStyle = die.fillStyle;
+            }
+            this.context?.beginPath();
+            this.context?.rect(
+                die.x,
+                die.y,
+                this.dimensions.width,
+                this.dimensions.height
             );
+            this.context?.fill();

```

Comment:
context isn't nullable in the type, can we remove all of these have optional chaining checks?